pkg/controllers/machine/disruption/emptiness.go

```diff
@@ -0,0 +1,92 @@
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	machineutil "github.com/aws/karpenter-core/pkg/utils/machine"
+)
+
+// Emptiness is a machine sub-controller that adds or removes status conditions on empty machines based on TTLSecondsAfterEmpty
+type Emptiness struct {
+	kubeClient client.Client
+	clock      clock.Clock
+}
+
+//nolint:gocyclo
+func (e *Emptiness) Reconcile(ctx context.Context, provisioner *v1alpha5.Provisioner, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	hasEmptyCondition := machine.StatusConditions().GetCondition(v1alpha5.MachineEmpty).IsTrue()
+	// From here there are three scenarios to handle:
+	// 1. If TTLSecondsAfterEmpty is not configured remove the emptiness status condition
+	if provisioner.Spec.TTLSecondsAfterEmpty == nil {
+		if hasEmptyCondition {
+			_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineEmpty)
+			logging.FromContext(ctx).Debugf("removing emptiness status condition from machine as emptiness has been disabled")
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Get the node to check the emptiness timestamp
+	node, err := machineutil.NodeForMachine(ctx, e.kubeClient, machine)
+	if err != nil {
+		if machineutil.IsDuplicateNodeError(err) || machineutil.IsNodeNotFoundError(err) {
+			_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineEmpty)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	emptinessTimestamp := node.Annotations[v1alpha5.EmptinessTimestampAnnotationKey]
+	if emptinessTimestamp == "" {
+		_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineEmpty)
+		return reconcile.Result{}, nil
+	}
+	emptyAt, err := time.Parse(time.RFC3339, emptinessTimestamp)
+	if err != nil {
+		_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineEmpty)
+		logging.FromContext(ctx).With("emptiness-timestamp", emptinessTimestamp).Errorf("unable to parse emptiness timestamp")
+		return reconcile.Result{}, nil //nolint:nilerr
+	}
+	emptinessTTLTime := emptyAt.Add(time.Duration(ptr.Int64Value(provisioner.Spec.TTLSecondsAfterEmpty)) * time.Second)
+
+	// 2. Otherwise, if the node is after emptiness expiration, but doesn't have the status condition, add it.
+	if e.clock.Now().After(emptinessTTLTime) && !hasEmptyCondition {
+		machine.StatusConditions().SetCondition(apis.Condition{
+			Type:     v1alpha5.MachineEmpty,
+			Status:   v1.ConditionTrue,
+			Reason:   "EmptinessTTLExceeded",
+			Severity: apis.ConditionSeverityWarning,
+		})
+		logging.FromContext(ctx).Debugf("marking machine as empty")
+		return reconcile.Result{}, nil
+	}
+	// 3. Finally, if the node isn't after emptiness expiration, but has the status condition, remove it.
+	if !e.clock.Now().After(emptinessTTLTime) && hasEmptyCondition {
```

Comment:
super nitty, but I think this should be incorrect - if the time is equal, we should return that it's expired. So here I think it should be `clock.Before(TTLTime)`, and then in line 76, it should be `!Before(TTLTime)` 

Although I think the likelihood that this happens and affects anyone is effectively zero, the ideology basically boils down Tennis vs Basketball on-the-line rules